File: BI-TJV/semestral/rest-server/src/test/java/wrzecond/service/TjvServiceTest.java
package wrzecond.service;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.mockito.BDDMockito;
import org.mockito.Mockito;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.dao.DataIntegrityViolationException;
import org.springframework.test.util.ReflectionTestUtils;
import wrzecond.allergen.TjvAllergenCreateDTO;
import wrzecond.allergen.TjvAllergenReadDTO;
import wrzecond.entity.TjvAllergen;
import wrzecond.repository.TjvAllergenRepository;

import javax.persistence.EntityNotFoundException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Random;

/**
 * This class tests TjvService.
 * As TjvService is generic template class,
 * I use TjvAllergenService for testing it.
 */
@SpringBootTest
public class TjvServiceTest {

    @Autowired
    private TjvAllergenService service;

    @MockBean
    private TjvAllergenRepository repository; // controlled by test

    @Test
    void getEntityByID () {
        TjvAllergen allergen = new TjvAllergen("Wheat");
        ReflectionTestUtils.setField(allergen, "id", (new Random()).nextInt(1000));

        // Mock definition
        BDDMockito.given(repository.getOne(allergen.getId())).willReturn(allergen);
        BDDMockito.given(repository.getOne(1000)).willThrow(EntityNotFoundException.class);

        // Test working
        Assertions.assertEquals(allergen, service.getEntityByID(allergen.getId()));
        BDDMockito.verify(repository, Mockito.atLeastOnce()).getOne(allergen.getId());

        // Test exception
        EntityNotFoundException exception = null;
        try { service.getEntityByID(1000); }
        catch (EntityNotFoundException x) { exception = x; }
        Assertions.assertNotNull(exception);
    }

    @Test
    void getByID () {
        TjvAllergen allergen = new TjvAllergen("Wheat");
        ReflectionTestUtils.setField(allergen, "id", (new Random()).nextInt(1000));
        TjvAllergenReadDTO dto = new TjvAllergenReadDTO(allergen.getId(), "Wheat");

        // Mock definition
        BDDMockito.given(repository.getOne(allergen.getId())).willReturn(allergen);
        BDDMockito.given(repository.getOne(1000)).willThrow(EntityNotFoundException.class);

        // Test working
        Assertions.assertEquals(dto, service.getByID(allergen.getId()));
        BDDMockito.verify(repository, Mockito.atLeastOnce()).getOne(allergen.getId());

        // Test exception
        EntityNotFoundException exception = null;
        try { service.getByID(1000); }
        catch (EntityNotFoundException x) { exception = x; }
        Assertions.assertNotNull(exception);
    }

    @Test
    void findByIDs () {
        TjvAllergen allergen = new TjvAllergen("Wheat"), allergen2 = new TjvAllergen("Milk");
        ReflectionTestUtils.setField(allergen, "id", (new Random()).nextInt(1000));
        ReflectionTestUtils.setField(allergen2, "id", 1000 + (new Random()).nextInt(1000));

        // Mock definition
        BDDMockito.given(repository.getOne(allergen.getId())).willReturn(allergen);
        BDDMockito.given(repository.getOne(allergen2.getId())).willReturn(allergen2);

        // Test working
        Assertions.assertEquals(service.findByIDs(Arrays.asList(allergen.getId(), allergen2.getId())), Arrays.asList(allergen, allergen2));
        BDDMockito.verify(repository, Mockito.atLeastOnce()).getOne(allergen.getId());
        BDDMockito.verify(repository, Mockito.atLeastOnce()).getOne(allergen2.getId());

        Assertions.assertEquals(service.findByIDs(Collections.singletonList(2000)), Collections.emptyList());
        BDDMockito.verify(repository, Mockito.atLeastOnce()).getOne(2000);
    }

    @Test
    void findAll () {
        // Test data
        TjvAllergen allergen = new TjvAllergen("Wheat"), allergen2 = new TjvAllergen("Milk");
        ReflectionTestUtils.setField(allergen, "id", (new Random()).nextInt(1000));
        ReflectionTestUtils.setField(allergen2, "id", 1000 + (new Random()).nextInt(1000));
        TjvAllergenReadDTO dto = new TjvAllergenReadDTO(allergen.getId(), "Wheat");
        TjvAllergenReadDTO dto2 = new TjvAllergenReadDTO(allergen2.getId(), "Milk");

        // Mock definition
        BDDMockito.given(repository.findAll(service.getSort())).willReturn(
            new ArrayList<TjvAllergen>() {{
                add(allergen); add(allergen2);
            }}
        );

        // Test working
        Assertions.assertEquals(service.findAll(), Arrays.asList(dto, dto2));
        BDDMockito.verify(repository, Mockito.atLeastOnce()).findAll(service.getSort());
    }

    @Test
    void create () {
        // Test data
        TjvAllergen allergen = new TjvAllergen("Wheat"), allergen2 = new TjvAllergen(null);
        TjvAllergenCreateDTO c_dto = new TjvAllergenCreateDTO("Wheat"),
                c_dto2 = new TjvAllergenCreateDTO(null);
        TjvAllergenReadDTO dto = new TjvAllergenReadDTO(allergen.getId(), "Wheat");

        // Mock definition
        BDDMockito.given(repository.save(allergen)).willReturn(allergen);
        BDDMockito.given(repository.save(allergen2)).willThrow(DataIntegrityViolationException.class);

        // Test working
        Assertions.assertEquals(dto, service.create(c_dto));
        BDDMockito.verify(repository, Mockito.atLeastOnce()).save(allergen);

        // Test exception
        DataIntegrityViolationException exception = null;
        try { service.create(c_dto2); }
        catch (DataIntegrityViolationException x) { exception = x; }

        Assertions.assertNotNull(exception);
    }

    @Test
    void createAll () {
        // Test data
        TjvAllergen allergen = new TjvAllergen("Wheat"), allergen2 = new TjvAllergen("Milk"),
            allergen3 = new TjvAllergen(null);
        TjvAllergenCreateDTO c_dto = new TjvAllergenCreateDTO("Wheat"),
                c_dto2 = new TjvAllergenCreateDTO("Milk"),
                c_dto3 = new TjvAllergenCreateDTO(null);
        TjvAllergenReadDTO dto = new TjvAllergenReadDTO(allergen.getId(), "Wheat"),
                dto2 = new TjvAllergenReadDTO(allergen2.getId(), "Milk");

        // Mock definition
        BDDMockito.given(repository.save(allergen)).willReturn(allergen);
        BDDMockito.given(repository.save(allergen2)).willReturn(allergen2);
        BDDMockito.given(repository.save(allergen3)).willThrow(DataIntegrityViolationException.class);

        // Test working
        Assertions.assertEquals(Arrays.asList(dto, dto2), service.create(Arrays.asList(c_dto, c_dto2)));
        BDDMockito.verify(repository, Mockito.atLeastOnce()).save(allergen);
        BDDMockito.verify(repository, Mockito.atLeastOnce()).save(allergen2);

        // Test exception
        DataIntegrityViolationException exception = null;
        try { service.create(Arrays.asList(c_dto2, c_dto3)); }
        catch (DataIntegrityViolationException x) { exception = x; }

        Assertions.assertNotNull(exception);
    }

    @Test
    void update () {
        TjvAllergen allergen = new TjvAllergen("Wheat"), allergen2 = new TjvAllergen(null);
        ReflectionTestUtils.setField(allergen, "id", (new Random()).nextInt(1000));
        ReflectionTestUtils.setField(allergen2, "id", 1000 + (new Random()).nextInt(1000));
        TjvAllergenCreateDTO c_dto = new TjvAllergenCreateDTO("Wheat"),
                c_dto2 = new TjvAllergenCreateDTO(null),
                c_dto3 = new TjvAllergenCreateDTO("Stone");

        // Mock definition
        BDDMockito.given(repository.getOne(allergen.getId())).willReturn(allergen);
        BDDMockito.given(repository.getOne(allergen2.getId())).willReturn(allergen2);
        BDDMockito.given(repository.getOne(2000)).willThrow(EntityNotFoundException.class);
        BDDMockito.given(repository.save(allergen2)).willThrow(DataIntegrityViolationException.class);

        // Test working
        service.update(allergen.getId(), c_dto);
        BDDMockito.verify(repository, Mockito.atLeastOnce()).getOne(allergen.getId());
        BDDMockito.verify(repository, Mockito.atLeastOnce()).save(allergen);

        // Test exception 1
        Exception exception = null;
        try { service.update(allergen2.getId(), c_dto2); }
        catch (DataIntegrityViolationException x) { exception = x; }
        Assertions.assertNotNull(exception);

        // Test exception 2
        exception = null;
        try { service.update(2000, c_dto3); }
        catch (EntityNotFoundException x) { exception = x; }
        Assertions.assertNotNull(exception);
    }

    @Test
    void delete () {
        // Mock definition
        BDDMockito.doNothing().when(repository).deleteById(1);
        BDDMockito.doThrow(EntityNotFoundException.class).when(repository).deleteById(2);

        // Test working
        service.delete(1);
        BDDMockito.verify(repository, Mockito.atLeastOnce()).deleteById(1);

        // Test exception
        EntityNotFoundException exception = null;
        try { service.delete(2); }
        catch (EntityNotFoundException x) { exception = x; }
        Assertions.assertNotNull(exception);
    }

}

File: BI-KOT/semestral/rest-server/src/test/kotlin/wrzecond/service/TjvServiceTest.kt
package wrzecond.service

import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.collections.shouldContainExactlyInAnyOrder
import io.kotest.matchers.shouldBe
import io.mockk.confirmVerified
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.springframework.dao.DataIntegrityViolationException
import org.springframework.data.domain.Sort
import org.springframework.web.server.ResponseStatusException
import wrzecond.TjvAllergenCreateDTO
import wrzecond.TjvAllergenReadDTO
import wrzecond.TjvAllergenUpdateDTO
import wrzecond.entity.TjvAllergen
import wrzecond.repository.TjvAllergenRepository
import javax.persistence.EntityNotFoundException

/**
 * This class tests TjvService.
 * As TjvService is generic template class,
 * I use TjvAllergenService for testing it.
 */
class TjvServiceTest : StringSpec ({
    // Init
    val mockRepository = mockk<TjvAllergenRepository>(relaxed = true)
    val service = TjvAllergenService(mockRepository)

    "getEntityByID" {
        val allergen = TjvAllergen("Wheat", 42)
        every { mockRepository.getOne(allergen.id) } returns allergen
        every { mockRepository.getOne(1000) } throws EntityNotFoundException()

        service.getEntityByID(allergen.id) shouldBe allergen
        verify { mockRepository.getOne(allergen.id) }
        confirmVerified(mockRepository)

        shouldThrow<ResponseStatusException> {
            service.getEntityByID(1000)
        }
        verify { mockRepository.getOne(1000) }
        confirmVerified(mockRepository)
    }
    "getByID" {
        val allergen = TjvAllergen("Milk", 1)
        val dto = TjvAllergenReadDTO(allergen.id, allergen.name)

        every { mockRepository.getOne(allergen.id) } returns allergen
        every { mockRepository.getOne(2000) } throws EntityNotFoundException()

        service.getByID(allergen.id) shouldBe dto
        verify { mockRepository.getOne(allergen.id) }
        confirmVerified(mockRepository)

        shouldThrow<ResponseStatusException> {
            service.getEntityByID(2000)
        }
        verify { mockRepository.getOne(2000) }
        confirmVerified(mockRepository)
    }
    "findByIDs" {
        val allergen1 = TjvAllergen("Cocoa", 10)
        val allergen2 = TjvAllergen("Beans", 12)

        every { mockRepository.getOne(allergen1.id) } returns allergen1
        every { mockRepository.getOne(allergen2.id) } returns allergen2

        service.findByIDs(listOf(allergen1.id, allergen2.id)) shouldContainExactlyInAnyOrder listOf(allergen1, allergen2)
        verify { mockRepository.getOne(allergen1.id) }
        verify { mockRepository.getOne(allergen2.id) }
        confirmVerified(mockRepository)
    }
    "findAll" {
        val allergen1 = TjvAllergen("Gluten", 24)
        val allergen2 = TjvAllergen("Cats", 25)
        val dto1 = TjvAllergenReadDTO(allergen1.id, allergen1.name)
        val dto2 = TjvAllergenReadDTO(allergen2.id, allergen2.name)

        every { mockRepository.findAll(any<Sort>()) } returns listOf(allergen1, allergen2)
        service.findAll() shouldContainExactlyInAnyOrder listOf(dto1, dto2)
        verify { mockRepository.findAll(any<Sort>()) }
        confirmVerified(mockRepository)
    }
    "create" {
        val allergen1 = TjvAllergen("School", 0)
        val allergen2 = TjvAllergen("C++", 0)
        val create1 = TjvAllergenCreateDTO(allergen1.name)
        val create2 = TjvAllergenCreateDTO(allergen2.name)
        val dto = TjvAllergenReadDTO(allergen1.id, allergen1.name)

        every { mockRepository.save(allergen1) } returns allergen1
        every { mockRepository.save(allergen2) } throws DataIntegrityViolationException("")

        service.create(create1) shouldContainExactlyInAnyOrder listOf(dto)
        verify { mockRepository.save(allergen1) }
        verify { mockRepository.flush() }
        confirmVerified(mockRepository)

        shouldThrow<ResponseStatusException> {
            service.create(create2)
        }
        verify { mockRepository.save(allergen2) }
        confirmVerified(mockRepository)
    }
    "update" {
        val allergen1 = TjvAllergen("Babish", 900)
        val allergen2 = TjvAllergen("Faltynek", 911)
        val allergen3 = TjvAllergen("Repka", 900)

        val update1 = TjvAllergenUpdateDTO("Repka")
        val update2 = TjvAllergenUpdateDTO(null)
        val update3 = TjvAllergenUpdateDTO("Stone")

        every { mockRepository.getOne(allergen1.id) } returns allergen1
        every { mockRepository.getOne(allergen2.id) } returns allergen2
        every { mockRepository.getOne(2000) } throws EntityNotFoundException()
        every { mockRepository.saveAndFlush(allergen3) } returns allergen3
        every { mockRepository.saveAndFlush(allergen2) } throws DataIntegrityViolationException("")

        service.update(allergen1.id, update1)
        verify { mockRepository.getOne(allergen1.id) }
        verify { mockRepository.saveAndFlush(allergen3) }
        confirmVerified(mockRepository)

        shouldThrow<ResponseStatusException> {
            service.update(allergen2.id, update2)
        }
        verify { mockRepository.getOne(allergen2.id) }
        verify { mockRepository.saveAndFlush(allergen2) }
        confirmVerified(mockRepository)

        shouldThrow<ResponseStatusException> {
            service.update(2000, update3)
        }
        verify { mockRepository.getOne(2000) }
        confirmVerified(mockRepository)
    }
    "delete" {
        every { mockRepository.deleteById(144) } returns Unit
        every { mockRepository.deleteById(142) } throws EntityNotFoundException()

        service.delete(144)
        verify { mockRepository.deleteById(144) }
        confirmVerified(mockRepository)

        shouldThrow<ResponseStatusException> {
            service.delete(142)
        }
        verify { mockRepository.deleteById(142) }
        confirmVerified(mockRepository)
    }

})